### ISSUE TITLE
Support for samtools split for non-string tags (e.g. haplotype HP:i)

### ISSUE THREAD

**Issue by ASLeonard** [samtools/samtools#1956]
I was hoping to use `samtools split -d HP` to split a phased bam file into its various haplotypes, but currently only string tags are supported. I tried to find a nicer solution, but this manual branch worked with existing htslib functions (pointers are *not* my friend) changing the [line here](https://github.com/samtools/samtools/blob/a4c73e4d41670957be6c68ff964a0310f58be9ab/bam_split.c#L691). Given there is already a `--max-split` option, was there a reason why integers are generally unsuitable compared to strings? I would imagine splitting by phase is not an obscure thing to do (and easier than multiple `samtools view -d HP:<phase>`). The other option would be to also encode phase/haplotype as a string, but would be redundant information.

```
char *val[2];
if (strcmp(tag, "HP")) sprintf(val,"%ld", bam_aux2i(tag));
else *val = tag ? bam_aux2Z(tag) : NULL;
```

I also noticed that if you split by tag, the format string is forcibly overwritten from whatever you provide. 

https://github.com/samtools/samtools/blob/a4c73e4d41670957be6c68ff964a0310f58be9ab/bam_split.c#L145-L147

The format string documentation definitely seems to indicate you can also change the `-f` format for TAGs.
```
Format string expansions:
  %%     %
  %*     basename
  %#     index (of @RG in the header, or count of TAG values seen so far)
  %!     @RG ID or TAG value
  %.     filename extension for output format
```

Best,
Alex

**Comment by nh13** [comment]
Different use case is to split the output of duplex reads from nanopore: https://github.com/nanoporetech/dorado?tab=readme-ov-file#duplex

**Comment by jkbonfield** [comment]
I thought we'd merged that already, but looking back it was `samtools view -d` (042d9db4e6ec4abb2cba39d83c2e7d258bf5fe36). 

https://github.com/samtools/samtools/pull/1934 added the -d option to split, so I agree it ought to follow the same logic as `view -d` does.

**Comment by daviesrob** [comment]
Now splitting on string tags works, it should be possible to extend to integer ones.  One question is how the numbers should be formatted?  Would it be better to go with the plain numeric value, or have leading zeroes, or leading zeroes and a `+/-` (although that might be a bad idea due to `+` being special in URLs...):

```
for i in 1 2 10 2147483647 ; do printf "%-11d %010d  %+011d\n%-11d %011d %+011d\n" $i $i $i -$i -$i -$i ; done
1           0000000001  +0000000001
-1          -0000000001 -0000000001
2           0000000002  +0000000002
-2          -0000000002 -0000000002
10          0000000010  +0000000010
-10         -0000000010 -0000000010
2147483647  2147483647  +2147483647
-2147483647 -2147483647 -2147483647
```

The idea behind zero padding would be that you can more easily sort them into order, although sadly [the unicode collation rules](http://www.unicode.org/reports/tr10/) mess up the negative values somewhat:
```
locale | grep COLLATE
LC_COLLATE="en_GB.utf8"
printf "%011d\n%010d\n%011d\n%010d\n%011d\n%010d\n%011d\n%010d\n" -1 1 -2 2 -10 10 -2147483647 2147483647 | sort
0000000001
-0000000001
0000000002
-0000000002
0000000010
-0000000010
2147483647
-2147483647
```

as opposed to:
```
printf "%011d\n%010d\n%011d\n%010d\n%011d\n%010d\n%011d\n%010d\n" -1 1 -2 2 -10 10 -2147483647 2147483647 | ( LC_ALL=C sort )
-0000000001
-0000000002
-0000000010
-2147483647
0000000001
0000000002
0000000010
2147483647
```

We could of course make this selectable by adding a new format code that prints strings as-is, but zero-pads numbers...

**Comment by ASLeonard** [comment]
You probably have to deal with a lot more use-cases (and I'm pretending binary representation of numbers is trivial), but I think plain numbers are better. Maybe samtools would allow this, but I doubt anyone would do `samtools view -d HP:+0000000001` if they wanted to just get `HP:i:1` records. If the bam record has `... HP:i:0` or `... dx:i:-1`, then the printed tags should probably just be "0" or "-1" respectively without any padding.

**Comment by daviesrob** [comment]
This has now been implemented via pull request #2005
